### ISSUE TITLE
Cleaning-alru parameters

### ISSUE THREAD

**Issue by khristenkovladimir** [Open-CAS/open-cas-linux#596]
## Question

Could you please help me to understand the "w" and "t" cleaning-alru parameters?
cache mode - wo.
By default, ("casadm -X -n cleaning-alru -i 1 -w 20 -s 120 -b 100 -t 10000") will dirty blocks be flushed if the time that has passed since the last write operation to block is more than 120 seconds and cache is idle (no I/O operation to cache instance) for more than 10 seconds?
So by default, if we have permanent I/O to cache instance, dirty blocks will not be flushed while the cache device is not fully filed. Am I correct?

I need the following behavior of cache - dirty blocks must be flushed if the time that passed since the last write operation to block is more than "s" parameter.

I have changed cleaning-alru parameters to - "casadm -X -n cleaning-alru -i 1 -w 20 -s 120 -b 100 -t 0" and then run the next tests:
First I write data in cache to 2 files:
 fio --name=cache --numjobs=8 --iodepth=4 --bs=4k --rw=randwrite --ioengine=libaio --direct=1 --group_reporting=1 --filename=/mnt/io1.dat:/mnt/io2.dat --runtime=60 --time_based=1
 
then immediately starting to read data from the fist file:
 fio --name=nvme --numjobs=8 --iodepth=4 --bs=4k --rw=randread --ioengine=libaio --direct=1 --group_reporting=1 --filename=/mnt/io1.dat --runtime=600 --time_based=1
 
but after 120 seconds, data from the second file has not been flushed to the core device. Why this has happened if -t 0?

Then I change parameters to - "casadm -X -n cleaning-alru -i 1 -w 0 -s 120 -b 100 -t 10000" and rerun the test.
And after 120 seconds, data from the second file started to flush to the core device. My cache device wasn't idle so why flush was started?

Thank you.

## Your Environment
OpenCAS version : 20.03.3
Operating System: CentOS 8.2
Kernel version: 4.18.0-193.28.1

**Comment by mmichal10** [comment]
Hi @khristenkovladimir, 

> By default, ("casadm -X -n cleaning-alru -i 1 -w 20 -s 120 -b 100 -t 10000") will dirty blocks be flushed if the time that has passed since the last write operation to block is more than 120 seconds and cache is idle (no I/O operation to cache instance) for more than 10 seconds?

By default the flush cleaning thread will wake up every 20s (`--wake-up` param) to check whether the time that has passed since the last write IO is greater than 120s (`--staleness-timethe` param) and whether the cache is idle longer than 10000ms (`--activity-threshold`). If the condition is fulfilled, the cleaning thread will start to flush dirty data in a 100 blocks chunks (`--flush-max-buffers` param) and will be flushing continuously until:
-  IO to the cache device was triggered,
- all the dirty blocks are flushed.

> I have changed cleaning-alru parameters to - "casadm -X -n cleaning-alru -i 1 -w 20 -s 120 -b 100 -t 0" and then run the next tests:
First I write data in cache to 2 files:
fio --name=cache --numjobs=8 --iodepth=4 --bs=4k --rw=randwrite --ioengine=libaio --direct=1 --group_reporting=1 --filename=/mnt/io1.dat:/mnt/io2.dat --runtime=60 --time_based=1
>
> then immediately starting to read data from the fist file:
> fio --name=nvme --numjobs=8 --iodepth=4 --bs=4k --rw=randread --ioengine=libaio --direct=1 --group_reporting=1 --filename=/mnt/io1.dat --runtime=600 --time_based=1
> 
> but after 120 seconds, data from the second file has not been flushed to the core device. Why this has happened if -t 0?

Has the flushing started at some later point? Or it didn't started at all?
Are you sure that there was no write requests? Maybe the filesystem you are using triggered writes? Could you modify your test and trigger IO to raw device? I tried to reproduce the behavior you described, and both on the raw device and the XFS filesystem flush occurred correctly. You can also observe a delay greater than time set with `-s 120` - the reason behind it is that the cleaning thread is being woken up every 20 seconds (`-w 20` param).

> 
> Then I change parameters to - "casadm -X -n cleaning-alru -i 1 -w 0 -s 120 -b 100 -t 10000" and rerun the test.
> And after 120 seconds, data from the second file started to flush to the core device. My cache device wasn't idle so why flush was started?
> 

I reproduced the behavior you are talking about using XFS filesystem, but with raw `/dev/cas1-1` device flushing didn't begin as long as cache wasn't idle for 10000ms. Since fio was reading a single file during the test, PageCache might cache it's content in RAM.

**Comment by khristenkovladimir** [comment]
Hi @mmichal10 

> Has the flushing started at some later point? Or it didn't started at all?
> Are you sure that there was no write requests? Maybe the filesystem you are using triggered writes? Could you modify your test and trigger IO to raw device? I tried to reproduce the behavior you described, and both on the raw device and the XFS filesystem flush occurred correctly. You can also observe a delay greater than time set with -s 120 - the reason behind it is that the cleaning thread is being woken up every 20 seconds (-w 20 param).

Flushing started only after the cache was idle. Yes, there were no write requests to the cache, only read requests.
I recorded a video for the xfs and lvm case:
xfs: https://www.dropbox.com/s/1516a3yn4l205xo/opencas_xfs_w20_s120_b100_t0.mp4?dl=0
lvm: https://www.dropbox.com/s/sbxb9c94ouvq7g9/opencas_lvm_w20_s120_b100_t0.mp4?dl=0

Does the --staleness-time option refer to the entire cache device or to individual blocks (default 4K) in the cache?

How can I achieve the following cache behavior - dirty blocks must be flushed if the time that passed since the last write operation to a block is more than "s" parameter and cache is not idle?

**Comment by khristenkovladimir** [comment]
@mmichal10 could you help me with this case?

Thank you.

**Comment by mmichal10** [comment]
Hi @khristenkovladimir 

> Does the --staleness-time option refer to the entire cache device or to individual blocks (default 4K) in the cache?

It refers to individual blocks

> How can I achieve the following cache behavior - dirty blocks must be flushed if the time that passed since the last write operation to a block is more than "s" parameter and cache is not idle?

You have to set both `activity-threshold` and `wake-up` parameters to 0

**Comment by khristenkovladimir** [comment]
Hi @mmichal10
Thank you for your reply.
I have updated my environment:
CentOS 8.2 to CentOS 8.3
OpenCAS 20.03.3 to OpenCAS 20.12.0.0438.master

I set the following parameters:
casadm -X -n cleaning-alru -i 1 -w 0 -s 30 -b 100 -t 0
 
And ran the following fio tests:
`fio --name=nvme --numjobs=6 --iodepth=4 --bs=4k --rw=randwrite --ioengine=libaio --direct=1 --group_reporting=1 --filename=/dev/cas1-1 --filesize=1G --runtime=600 --time_based=1`

So all blocks are rewritten every few seconds, and I have expected that dirty blocks will be flushed to the core device after the fio test finishes + the 30 seconds (--staleness-time parameter). 
But after the 30 seconds, after the fio test started, the dirty blocks began periodically flushing to the core device.

Why did it happen?

I have recorded the video:
https://user-images.githubusercontent.com/11570512/106269640-71627a00-6235-11eb-8ed4-dfad390f182a.mp4

**Comment by mmichal10** [comment]
@khristenkovladimir 

The behavior with dirty blocks being flushed 30 seconds after write is a bug. We are preparing a fix. 

I will let you know as soon as it will be fixed

**Comment by robertbaldyga** [comment]
Hi @khristenkovladimir,

FYI fix for this issue is already merged to our master branch.

It has been fixed in this PR https://github.com/Open-CAS/open-cas-linux/pull/700.
It contains update of OCF submodule, where the actual fix (https://github.com/Open-CAS/ocf/pull/458) has been implemented.

Please let us know if it solves your problem.

**Comment by khristenkovladimir** [comment]
Hi @robertbaldyga, @mmichal10

I had checked and all works perfectly.

Thank you very much!

**Comment by mslovy** [comment]
@robertbaldyga  @mmichal10 , I wonder if any trade-off to setting  -w 0 and -t 0 . Does Frequently flushing dirty impact write/read cache performance？

Assume that HDD and NVME SSD is not the IO bottleneck, I want to know if the client IO will be blocked during flush dirty data because of locks. And If so the lock is fine-grain for each cache block or all cache blocks in a cache set?

**Comment by mmichal10** [comment]
Hi @mslovy 

If you have any questions I strongly recommend posting it a separate issue ([and labeling it as `question`](https://github.com/Open-CAS/open-cas-linux/issues?q=is%3Aissue+is%3Aopen+label%3Aquestion)). We hardly ever review resolved issues.
